frontend_tests/node_tests/narrow_activate.js

```diff
@@ -0,0 +1,201 @@
+        target_list.add_messages(messages, opts);
+    };
+
+    // We have strange hacks in narrow.activate to sleep 0
+    // seconds.
+    global.patch_builtin('setTimeout', (f, t) => {
+        assert.equal(t, 0);
+        f();
+    });
+
+    return {
+        clear: () => {
+            events = [];
+        },
+        push_event: (event) => {
+            events.push(event);
+        },
+        assert_events: (expected_events) => {
+            assert.deepEqual(expected_events, events);
+        },
+    };
+}
+
+function stub_message_list() {
+    message_list.MessageList = function (table_name, filter) {
+        this.messages = [];
+        this.filter = filter;
+
+        return this;
+    };
+
+    message_list.MessageList.prototype = {
+        add_messages: function (messages) {
+            var predicate = this.filter.predicate();
+            messages = _.filter(messages, predicate);
+            this.messages = this.messages.concat(messages);
+        },
+
+        empty: function () {
+            return this.messages.length === 0;
+        },
+
+        first_unread_message_id: function () {
+            return this.messages[0].id;
+        },
+
+        select_id: function (msg_id) {
+            this.selected_id = msg_id;
+        },
+    };
+}
+
+(function test_basics() {
+    stub_message_list();
+
+    var helper = test_helper();
+    var terms = [
+        { operator: 'stream', operand: 'Denmark' },
+    ];
+    var opts = {
+        select_first_unread: true,
+    };
+
+    var selected_id = 1000;
+
+    var selected_message = {
+        id: selected_id,
+        type: 'stream',
+        stream_id: denmark.stream_id,
+    };
+
+    var messages = [selected_message];
+
+    home_msg_list.selected_id = () => { return selected_id; };
+    current_msg_list.selected_id = () => { return -1; };
+
+    message_list.all = {
+        all_messages: () => {
+            return messages;
+        },
+        get: (msg_id) => {
+            assert.equal(msg_id, selected_id);
+            return selected_message;
+        },
+    };
+
+    var cont;
+
+    message_fetch.load_messages_for_narrow = (opts) => {
+        cont = opts.cont;
+
+        assert.deepEqual(opts, {
+            cont: opts.cont,
+            then_select_id: selected_id,
+            use_initial_narrow_pointer: false,
+        });
+    };
+
+    narrow.activate(terms, opts);
+
+    assert.equal(message_list.narrowed.selected_id, selected_id);
+
+    helper.assert_events([
+        'notifications.clear_compose_notifications',
+        'notifications.redraw_title',
+        'ui_util.change_tab_to',
+        'message_scroll.hide_indicators',
+        'unread_ops.process_visible',
+        'hashchange.save_narrow',
+        'search.update_button_visibility',
+        'compose_actions.on_narrow',
+        'top_left_corner.handle_narrow_activated',
+        'stream_list.handle_narrow_activated',
+        'trigger event',
+    ]);
+
+    channel.post = (opts) => {
```

Comment:
I'm interested to why this is not earlier because there seems to be another `maybe_report_narrow_time` in the `setTimeout` (L257-260). Seems like that is also run, so we can assert it in the helper above.